csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/Assets.cs

```diff
@@ -0,0 +1,162 @@
+        ///     }
+        ///   }
+        /// }
+        /// 
+        /// Returns dependencies
+        ///   Required = {
+        ///     "castle.core/4.4.1/lib/netstandard1.5/Castle.Core.dll",
+        ///     "json.net/1.0.33/lib/netstandard2.0/Json.Net.dll"
+        ///   }
+        ///   UsedPackages = {
+        ///     "castle.core",
+        ///     "json.net"
+        ///   }
+        /// </summary>
+        private Dependencies AddPackageDependencies(JObject json, Dependencies dependencies)
+        {
+            // If there are more than one framework we need to pick just one.
+            // To ensure stability we pick one based on the lexicographic order of
+            // the framework names.
+            var references = json
+                .GetProperty("targets")?
+                .Properties()?
+                .MaxBy(p => p.Name)?
+                .Value
+                .ToObject<Dictionary<string, ReferenceInfo>>();
+
+            if (references is null)
+            {
+                progressMonitor.LogDebug("No references found in the targets section in the assets file.");
+                return dependencies;
+            }
+
+            // Find all the compile dependencies for each reference and
+            // create the relative path to the dependency.
+            return references
+                .Aggregate(dependencies, (deps, r) =>
+                {
+                    var info = r.Value;
+                    var name = r.Key.ToLowerInvariant();
+                    if (info.Type != "package")
+                    {
+                        return deps;
+                    }
+
+                    // If this is a .NET framework reference then include everything.
+                    return netFrameworks.Any(framework => name.StartsWith(framework))
+                        ? deps.Add(name, "")
+                        : info
+                            .Compile
+                            .Aggregate(deps, (d, p) => d.Add(name, p.Key));
+                });
+        }
+
+        /// <summary>
+        /// Parse `json` as project.assets.json content and add relative paths to the dependencies
+        /// (together with used package information) required for compilation.
+        /// </summary>
+        /// <returns>True if parsing succeeds, otherwise false.</returns>
+        public bool TryParse(string json, Dependencies dependencies)
+        {
+            try
+            {
+                var obj = JObject.Parse(json);
+                AddPackageDependencies(obj, dependencies);
+                return true;
+            }
+            catch (Exception e)
+            {
+                progressMonitor.LogDebug($"Failed to parse assets file (unexpected error): {e.Message}");
+                return false;
+            }
+        }
+
+        public static Dependencies GetCompilationDependencies(ProgressMonitor progressMonitor, IEnumerable<string> assets)
```

Comment:
I think you could remove the `ProgressMonitor progressMonitor` parameter if the method was not `static`.

Reply:
`GetCompilationDependencies` is a factory like method (it creates the Assets object which needs the progress monitor). I chose to do it like this because we only need the Assets object once.

Reply:
Ahh, sorry, I missed that this is a `public` factory method.